### PR TITLE
driver/at86rf2xx: add an option for auto-ack with the pending bit

### PR DESCRIPTION
**Pull request by Hyungsin**
### Contribution description

IEEE 802.15.4 provides ACK with the pending bit for data polling mechanism.
This PR adds this option to the at86rf2xx driver.

### CODE REVIEW

**Comment by gebart** [comment]
How do you use this in practice? Is it controlled by the application, or by the network stack automatically? What does the usage scenario look like?
I think this is an interesting feature, but I don't understand where it goes in the big picture.

**Comment by Hyungsin** [comment]
@gebart, This feature is necessary for operating Openthread leaf nodes. 

1. An openthread leaf node periodically wakes up and sends a data request packet to its parent node.
2. The parent node sends an ACK with (or without) the pending bit if it has packets to send to the leaf node (or not).
3. The leaf node wakes up for a while to receive packets from the parent node if it receives an ACK with the pending bit. Otherwise, it sleeps immediately.

File: drivers/include/at86rf2xx.h [hunk 1]
```diff
@@ -135,6 +135,7 @@ extern "C" {
 #define AT86RF2XX_OPT_SRC_ADDR_LONG  (NETDEV_IEEE802154_SRC_MODE_LONG)  /**< legacy define */
 #define AT86RF2XX_OPT_RAWDUMP        (NETDEV_IEEE802154_RAW)            /**< legacy define */
 #define AT86RF2XX_OPT_AUTOACK        (NETDEV_IEEE802154_ACK_REQ)        /**< legacy define */
+#define AT86RF2XX_OPT_ACK_PENDING    (NETDEV_IEEE802154_FRAME_PEND)     /**< legacy define */
```
Comment (jia200x) [inline drivers/include/at86rf2xx.h]: This line fail to compile. There's no NETDEV_IEEE802154_FRAME_PEND directive in this branch.

Comment (jia200x) [inline drivers/include/at86rf2xx.h]: `git add drivers/include/net/netdev/ieee802154.h`?

Comment (Hyungsin) [reply drivers/include/at86rf2xx.h]: Exactly. Sorry for the mistake! :)

**Comment by Hyungsin** [comment]
@jia200x, addressed your comment. Thanks!

File: drivers/include/at86rf2xx.h (hunk 1 above)
Comment (jia200x) [reply drivers/include/at86rf2xx.h]: NP ;)

**Comment by Hyungsin** [comment]
@jia200x, Any further comment or is it good?

File: drivers/at86rf2xx/at86rf2xx_getset.c [hunk 2]
```diff
@@ -380,6 +380,13 @@ void at86rf2xx_set_option(at86rf2xx_t *dev, uint16_t option, bool state)
                           : (tmp & ~AT86RF2XX_IRQ_STATUS_MASK__RX_START);
             at86rf2xx_reg_write(dev, AT86RF2XX_REG__IRQ_MASK, tmp);
             break;
+        case AT86RF2XX_OPT_ACK_PENDING:
+            DEBUG("[at86rf2xx] opt: enabling pending ACKs\n");
+            tmp = at86rf2xx_reg_read(dev, AT86RF2XX_REG__CSMA_SEED_1);
+            tmp = (state) ? (tmp |= AT86RF2XX_CSMA_SEED_1__AACK_SET_PD)
```
Comment (jia200x) [inline drivers/at86rf2xx/at86rf2xx_getset.c]: this lines gives undefined behavior. Please change the `|=` to `|`

**Comment by jia200x** [comment]
Just requested a minor change, but code looks good to me.

File: sys/include/net/netopt.h [hunk 3]
```diff
@@ -147,6 +147,8 @@ typedef enum {
                                  *   the current state */
     NETOPT_AUTOACK,             /**< en/disable link layer auto ACKs or read
                                  *   the current state */
+    NETOPT_ACK_PENDING,         /**< en/disable link layer auto ACKs with
```
Comment (jia200x) [inline sys/include/net/netopt.h]: Could you refactor a little bit this comments? Just to make the feature more clear

**Comment by Hyungsin** [comment]
@jia200x, fixed again. Check please :)

**Comment by gebart** [comment]
How does the OpenThread router node know when to configure the pending bit in advance if there are more than one leaf node connected to it? I mean since this flag must be set before the frame is received, how does it know that the particular leaf node that has some traffic waiting will be asking before the other leaf nodes do?

**Comment by Hyungsin** [comment]
@gebart, OpenThread has a packet destination list to support this feature. The current RIOT-OpenThread port does not support this. But my OpenThread port includes this feature. I will open a new PR  after #7149 is merged, which takes much longer than expected though...

**Comment by gebart** [comment]
@Hyungsin I don't think that answers my question though. How does the OpenThread router know in advance which child node it will be receiving a frame from, to know when to set or not set the pending bit? Or is the pending bit set on all ACKs whenever there is something to send to any node, regardless of whether the actual destination node is the one sending the query?

**Comment by jia200x** [comment]
@gebart from the AT86RF datashet:

> The  content  of  AACK_SET_PD  bit  is  copied  into  the  frame  pending  subfield  of  the  
acknowledgment  frame  if  the  ACK  is  the  response  to  a  data  request  MAC  command  
frame.

Also in the OpenThread SAMR21 radio implementation, the SrcMatch feature is not implemented.
So, it seems it will reply to every data request command, regardless of the source address. @Hyungsin can you confirm this?

[the SrcMatch feature is not implemented]:https://github.com/openthread/openthread/blob/e1ce2fb35d90c7893cc8a24624d9082793e94a5d/examples/platforms/samr21/radio.c#L575

**Comment by jia200x** [comment]
@Hyungsin could you maybe open a PR with the new features you are working on? It would be very nice to see these features and #7149 shouldn't change so much until gets merged.

**Comment by jia200x** [comment]
@Hyungsin ok, code looks good to me. 

@PeterKietzmann could you have a quick look too?

**Comment by Hyungsin** [comment]
@jia200x, @gebart, yes what @jia200x said is right.
So with at86rf233 radio, the SrcMatch feature is expected to set AACK_SET_PD if OpenThread has at least one pending data, which makes the radio send an ACK to every data request command it receives. If OpenThread has no pending data, it disable AACK_SET_PD.

**Comment by Hyungsin** [comment]
@jia200x, sure, let me open the PR soon :)

File: sys/include/net/netopt.h [hunk 4]
```diff
@@ -147,6 +147,12 @@ typedef enum {
                                  *   the current state */
     NETOPT_AUTOACK,             /**< en/disable link layer auto ACKs or read
                                  *   the current state */
+    NETOPT_ACK_PENDING,         /**< en/disable the frame pending bit of ACKs.
```
Comment (PeterKietzmann) [inline sys/include/net/netopt.h]: Please also add this option [here](https://github.com/RIOT-OS/RIOT/blob/d15c7559f342985a64c969f8e6d93eb650a8503e/sys/net/crosslayer/netopt/netopt.c#L26).

**Comment by PeterKietzmann** [comment]
Murock failed due to the missing string mapping in netopt.c

File: sys/include/net/netopt.h (hunk 4 above)
Comment (Hyungsin) [reply sys/include/net/netopt.h]: Done. Thanks!

**Comment by Hyungsin** [comment]
@PeterKietzmann, @jia200x, now all checks are passed :)

**Comment by PeterKietzmann** [comment]
@gebart @jia200x any objections to merge this PR?

**Comment by gebart** [comment]
@PeterKietzmann no objections, only the same concerns that you raised above to not add any unused functionality, but if the OpenThread stack uses this feature I say let's add it.

**Comment by jia200x** [comment]
@PeterKietzmann in fact, Sleepy End Devices in OpenThread are possible with this PR.
The current mainstream OT is incomplete without oending bit management

**Comment by Hyungsin** [comment]
@gebart, Done! 👍

**Comment by jia200x** [comment]
@gebart Do you have additional comments on this PR?

**Comment by jia200x** [comment]
> Nothing more to comment on

Ok, great. I will merge.
Thanks @Hyungsin for your contribution.

**Pull request merged**